In [ ]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/rerkki/data-analytics-ml/refs/heads/main/datasets/concrete.csv", encoding='latin-1')
df

#Cement : measured in kg in a m3 mixture
#Blast : measured in kg in a m3 mixture
#Fly ash : measured in kg in a m3 mixture
#Water : measured in kg in a m3 mixture
#Superplasticizer : measured in kg in a m3 mixture
#Coarse Aggregate : measured in kg in a m3 mixture
#Fine Aggregate : measured in kg in a m3 mixture
#Age : day (1~365)
#Concrete compressive strength measured in MPa

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import numpy as np

# Define features (X) and target (y)
X = df.drop('strength', axis=1)
y = df['strength']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse) # Fixed: Calculate RMSE using np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error (MSE): {mse:.2f}')
print(f'Root Mean Squared Error (RMSE): {rmse:.2f}')
print(f'Mean Absolute Error (MAE): {mae:.2f}')
print(f'R-squared (R2): {r2:.2f}')

In [ ]:
import pandas as pd

# Display the coefficients of the model
coefficients_df = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
display(coefficients_df)

print(f'Intercept: {model.intercept_:.2f}')

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Visualization of Coefficients
plt.figure(figsize=(10, 6))
sns.barplot(x='Coefficient', y='Feature', data=coefficients_df, hue='Feature', palette='viridis', legend=False) # Fixed FutureWarning
plt.title('Feature Coefficients in Linear Regression Model')
plt.xlabel('Coefficient Value')
plt.ylabel('Feature')
plt.grid(axis='x', linestyle='--', alpha=0.7)
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Calculate the correlation matrix
correlation_matrix = df.corr()

# Plotting the correlation heatmap
plt.figure(figsize=(10, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt='.2f', linewidths=.5)
plt.title('Correlation Map of All Variables')
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Visualization of Predicted vs Actual Strength
plt.figure(figsize=(8, 8))
sns.scatterplot(x=y_test, y=y_pred, alpha=0.7)
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], 'r--', lw=2) # Ideal prediction line
plt.title('Predicted vs. Actual Concrete Strength')
plt.xlabel('Actual Strength')
plt.ylabel('Predicted Strength')
plt.grid(True, linestyle='--', alpha=0.7)
plt.show()

# Task
This exercise aims to provide a comprehensive understanding of multilinear regression using the provided concrete strength dataset. You will train a model, interpret its various components, and analyze the relationships between different features and the target variable.

**Task 1: Model Training and Evaluation Parameter Tuning**

Experiment with the `test_size` parameter in the `train_test_split` function. Try values like `0.1`, `0.3`, and `0.5`. Additionally, experiment with the `random_state` parameter, trying values like `0`, `10`, and `100`. For each combination of `test_size` and `random_state`, train the linear regression model and record the evaluation metrics (Mean Squared Error (MSE), Root Mean Squared Error (RMSE), Mean Absolute Error (MAE), and R-squared (R2)).

After running these experiments, discuss the following:
*   How do changes in `test_size` affect the evaluation metrics? Explain why different split ratios might lead to variations in model performance and generalization.
*   How does changing the `random_state` influence the metrics, especially when `test_size` is kept constant? What does this tell you about the randomness in data splitting?
*   Based on your observations, what are the implications of choosing different `test_size` and `random_state` values for model evaluation?

## Task 1: Model Training and Evaluation Parameter Tuning

### Subtask:
Experiment with different `test_size` and `random_state` values in the `train_test_split` function, train a Linear Regression model for each combination, record evaluation metrics, and discuss the impact of these parameters.


**Reasoning**:
The first step is to define the `test_size` and `random_state` values as instructed and initialize an empty list to store the experimental results. This sets up the necessary variables before starting the iterative process.



In [7]:
test_sizes = [0.1, 0.3, 0.5]
random_states = [0, 10, 100]
results = []

print("Defined test_sizes and random_states, and initialized results list.")

Defined test_sizes and random_states, and initialized results list.


**Reasoning**:
Now that the parameters are initialized, the next step is to iterate through each combination of `test_size` and `random_state`, perform the `train_test_split`, train a Linear Regression model, make predictions, calculate the specified evaluation metrics (MSE, RMSE, MAE, R2), and store these results in the `results` list. After the iterations, the results will be converted into a Pandas DataFrame for structured analysis.



In [8]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import numpy as np
import pandas as pd

# Ensure X and y are defined from previous steps
# X = df.drop('strength', axis=1)
# y = df['strength']

for ts in test_sizes:
    for rs in random_states:
        # a. Split the X and y data into training and testing sets
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=ts, random_state=rs)

        # b. Initialize a new LinearRegression model
        model = LinearRegression()

        # c. Train the model using the training data
        model.fit(X_train, y_train)

        # d. Make predictions on the test set
        y_pred = model.predict(X_test)

        # e. Calculate the evaluation metrics
        mse = mean_squared_error(y_test, y_pred)
        rmse = np.sqrt(mse)
        mae = mean_absolute_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)

        # f. Store the results
        results.append({
            'test_size': ts,
            'random_state': rs,
            'MSE': mse,
            'RMSE': rmse,
            'MAE': mae,
            'R2': r2
        })

# Convert the results list into a Pandas DataFrame
results_df = pd.DataFrame(results)
print("Experiments completed and results stored in 'results_df'.")
display(results_df)

Experiments completed and results stored in 'results_df'.


,test_size,random_state,MSE,RMSE,MAE,R2
0,0.1,0,112.152385,10.590202,8.076963,0.602721
1,0.1,10,116.589096,10.797643,8.536279,0.535495
2,0.1,100,105.138542,10.253709,8.268165,0.635333
3,0.3,0,112.130648,10.589176,8.180855,0.584797
4,0.3,10,126.445426,11.244796,8.883646,0.521183
5,0.3,100,100.913526,10.045572,8.036753,0.621782
6,0.5,0,120.133851,10.960559,8.642854,0.572574
7,0.5,10,117.996570,10.862623,8.558158,0.581393
8,0.5,100,106.151654,10.302992,8.095344,0.582509


### Discussion of Model Training and Evaluation Parameter Tuning

Based on the `results_df` obtained from experimenting with different `test_size` and `random_state` values, we can analyze the impact of these parameters on the model's evaluation metrics:

**1. How changes in `test_size` affect the evaluation metrics and why different split ratios might lead to variations in model performance and generalization:**

*   **General Trend**: As `test_size` increases (from 0.1 to 0.5), we generally observe variations in MSE, RMSE, MAE, and R2. There isn't a consistent linear trend of improvement or degradation across all metrics with increasing `test_size`, as the `random_state` also plays a significant role.
*   **Small `test_size` (e.g., 0.1)**: With a smaller test set, the evaluation metrics might be less representative of the model's true performance on unseen data. The model is trained on a larger portion of the data, which could lead to better training performance but the test set might not be diverse enough to accurately reflect generalization. For instance, `random_state=100` with `test_size=0.1` yielded the best R2 (0.635) and lowest RMSE (10.254), suggesting a potentially 'lucky' split where the small test set was well-predicted.
*   **Medium `test_size` (e.g., 0.3)**: This seems to offer a more balanced approach between training data size and test data representativeness. The metrics here show a range of performance, with `random_state=100` again yielding strong results (R2=0.622, RMSE=10.046), suggesting that with a larger test set, the model's performance can still be robust.
*   **Large `test_size` (e.g., 0.5)**: When `test_size` is 0.5, half of the data is used for testing. This leaves less data for training, which could potentially lead to a less robust model as it has seen fewer examples during training. However, the larger test set provides a more comprehensive evaluation of the model's generalization capabilities. For example, `test_size=0.5` consistently shows a slightly higher RMSE and lower R2 compared to the best `test_size=0.3` case (e.g., RMSE for `test_size=0.5` ranges from 10.303 to 10.961, while for `test_size=0.3` it goes down to 10.046). This might indicate that the reduced training data limited the model's ability to learn complex patterns.
*   **Reason for Variations**: Different split ratios affect the amount and distribution of data available for training and testing. A smaller test set might by chance contain easier-to-predict samples, leading to overly optimistic metrics. Conversely, a larger test set might contain a broader range of data, providing a more realistic but potentially less flattering view of performance, especially if the training set was too small to capture all necessary patterns.

**2. How changing the `random_state` influences the metrics, especially when `test_size` is kept constant, and what this tells you about the randomness in data splitting:**

*   **Significant Impact**: Across all `test_size` values, changing `random_state` clearly influences all evaluation metrics. For example, with `test_size=0.1`, R2 varies from 0.535 (random_state=10) to 0.635 (random_state=100). Similarly, with `test_size=0.3`, R2 ranges from 0.521 (random_state=10) to 0.622 (random_state=100). This variation highlights the stochastic nature of data splitting.
*   **Randomness in Data Splitting**: The `random_state` parameter controls the shuffling applied to the data before splitting. Different `random_state` values lead to different subsets of data being allocated to the training and testing sets. This means the specific samples that constitute the test set vary, and the difficulty of predicting those samples also varies.
*   **Impact on Model Evaluation**: A particularly 'easy' or 'difficult' test set, due to a specific random split, can lead to seemingly good or bad performance metrics, respectively. This demonstrates that a single run with a specific `random_state` might not give a stable or truly representative measure of the model's performance. It's crucial to acknowledge that the chosen `random_state` can introduce bias into the evaluation if not handled carefully.

**3. Implications of choosing different `test_size` and `random_state` values for model evaluation:**

*   **`test_size` Implications**:
    *   **Bias-Variance Trade-off**: A very small `test_size` can lead to high variance in evaluation metrics (the metrics might vary widely if the split changes slightly), while a very large `test_size` reduces the training data, potentially leading to a model with higher bias (underfitting).
    *   **Representativeness**: The test set should ideally be representative of the overall data distribution. A `test_size` that is too small might not achieve this, leading to unreliable performance estimates.
    *   **Common Practice**: A `test_size` between 0.2 and 0.3 is often a good compromise, providing enough data for training while reserving a sufficiently large and representative set for testing.

*   **`random_state` Implications**:
    *   **Reproducibility**: Setting a fixed `random_state` is essential for reproducibility of experiments. Without it, running the same code multiple times would yield different results, making comparisons difficult.
    *   **Robustness**: Relying on a single `random_state` can lead to an evaluation that is specific to that particular split. To get a more robust estimate of model performance, it's advisable to perform cross-validation (e.g., K-fold cross-validation) or run the train-test split multiple times with different `random_state` values and average the results. This helps mitigate the impact of a 'lucky' or 'unlucky' split.

**Conclusion**: The experiments clearly illustrate that both `test_size` and `random_state` are critical hyper-parameters for model evaluation. Choosing appropriate values and understanding their impact is essential for obtaining reliable and generalizable performance metrics. For robust evaluation, it's often recommended to use cross-validation or repeat the train-test split process with different random seeds to average out the variability introduced by data partitioning.